### PR TITLE
feat: vortex.dataset.Dataset: deep integration with Polars & DuckDB

### PR DESCRIPTION
**Pull request by danking**
1. I added a RecordBatchReader to vortex-serde to support DuckDB.
2. I made Python logging only show Info and higher (the compressor is too noisy).
3. I added vortex._lib.Dataset and vortex.dataset.Dataset which work together to implement the [arrow.dataset.Dataset](https://arrow.apache.org/docs/python/api/dataset.html) API.
4. (driveby) Cleanup docs for `Array.to_numpy`.
5. I wrote a Substrait Expression to Vortex Expression translator and used that to implement Arrow to Vortex.

### CODE REVIEW

File: pyvortex/src/dataset.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            .build()?
+            .block_on(io::layout_reader(
+                self_.fname(),
+                projection,
+                batch_size,
+                row_filter,
+            ))
+            .map_err(PyVortexError::map_err)?;
+
+        let record_batch_rader: Box<dyn RecordBatchReader + Send> =
```
Comment (lwwmanning) [inline pyvortex/src/dataset.rs]: nit: rader

File: vortex-serde/src/layouts/read/recordbatchreader.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+fn vortex_to_arrow(result: VortexResult<Array>) -> Result<RecordBatch, ArrowError> {
+    result
+        .and_then(RecordBatch::try_from)
+        .map_err(vortex_to_arrow_error)
+}
+
+pub struct VortexRecordBatchReader<R: VortexReadAt + Unpin + Send + 'static> {
+    stream: LayoutBatchStream<R>,
+    arrow_schema: SchemaRef,
+    runtime: tokio::runtime::Runtime,
```
Comment (lwwmanning) [inline vortex-serde/src/layouts/read/recordbatchreader.rs]: 👀

Comment (lwwmanning) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: this is sus, tagging @robert3005 because I don't have an immediately better suggestion

Comment (AdamGS) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: I think it makes sense to have the runtime as a static thing, its not really bound to the lifetime of the reader, and I suspect you might run into weird issues if you have multiple runtimes.

Comment (AdamGS) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: I recently ran into [this](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#current-thread-scheduler) note in Tokio's docs, the way I read it all readers spawned from the same thread share underlying resources, which essentially makes the IO/serialization serialized between them, which is really bad.

Comment (robert3005) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: Agree with @AdamGS we should have a lazy_static in the library with the runtime that all python functions can use

File: vortex-serde/src/layouts/read/recordbatchreader.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+impl<R: VortexReadAt + Unpin + Send + 'static> VortexRecordBatchReader<R> {
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/recordbatchreader.rs]: fwiw this really only needs Unpin and 'static you can kill the Send bound all throughout

File: vortex-serde/src/layouts/read/recordbatchreader.rs (hunk 2 above)
Comment (danking) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: Done.

File: vortex-serde/src/layouts/read/recordbatchreader.rs (hunk 3 above)
Comment (danking) [reply vortex-serde/src/layouts/read/recordbatchreader.rs]: Hmm. Indeed. I got pretty confused at one point trying to create a `Box<dyn RecordBatchReader + Send>`. Removed.

File: pyvortex/src/dataset.rs (hunk 1 above)
Comment (danking) [reply pyvortex/src/dataset.rs]: Done.

**Comment by danking** [comment]
@lwwmanning I parameterized the reader by a Runtime

**Comment by danking** [comment]
Benchmarks reverted to current thread runtime.

File: pyvortex/src/lib.rs [hunk 4]
```diff
@@ -2,21 +2,41 @@
...
 mod io;
 mod python_repr;
+use lazy_static::lazy_static;
+use log::LevelFilter;
+use pyo3_log::{Caching, Logger};
+use tokio::runtime::Runtime;
+use vortex_error::{VortexError, VortexExpect as _};
+
+lazy_static! {
+    static ref TOKIO_RUNTIME: Runtime = Runtime::new()
```
Comment (lwwmanning) [inline pyvortex/src/lib.rs:26]: I *think* this is fine? though tagging @robert3005 @AdamGS who are more familiar with potential weirdness around python/async stuff

**Review by lwwmanning — APPROVED** [review]

File: pyvortex/python/vortex/dataset.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        columns: list[str] | None = None,
+        filter: pc.Expression | None = None,
+        batch_size: int | None = None,
+        batch_readahead: int | None = None,
+        fragment_readahead: int | None = None,
+        fragment_scan_options: pa.dataset.FragmentScanOptions | None = None,
+        use_threads: bool | None = None,
+        memory_pool: pa.MemoryPool = None,
+    ) -> pa.dataset.Scanner:
+        """Not implemented."""
```
Comment (robert3005) [inline pyvortex/python/vortex/dataset.py:106]: Is this truly not implemented or leftover?

Comment (danking) [reply pyvortex/python/vortex/dataset.py:106]: Indeed a mistake; already fixed in develop: https://github.com/spiraldb/vortex/commit/54c19938d1b2dcff5bdcdcca673f353f9a9d9610#diff-482b5dfc27f9f86d89ce96b259d2781464eb749b777c79fe2dca8231e6f26983L117

**Pull request merged**